linode/manager
# fix: [M3-9022, M3-9027] - Filter regions by endpoints and fix unavailable regions

## Description 📝
1. Warning notices for unavailable Gen2 regions are not displaying the region names due to inconsistencies in error response formats.
2. Regions displayed in the Object Gen2 create drawer and Access Keys list are not being filtered based on user endpoint capabilities.

> [!note]
This code relies on the `/endpoints`, which is activated through customer tags and capabilities. If endpoints is not available, we fallback to using regions and clusters for bucket retrieval.

## Changes  🔄
- Regions themselves have capabilities, therefore we need to filter those by what is returned from the `object-storage/endpoints` endpoint. This has to be done in 3 places:
    1. The create bucket drawer when selecting a region
    2. The access key `/object-storage/access-keys` page when displaying regions
    3. The access key drawer when selecting "Show All"     
- For the unavailable regions, instead of passing the entire region to the component we only need the labels. This helped with some of the type differences.
- New `useObjectStorageRegions` hook that filters regions by `endpoints` if user has Gen2 capabilities, otherwise falls back to all regions for backwards compatibility for legacy and multi-cluster.

## Target release date 🗓️
Before 12/20 (Friday)

## Preview 📷
| Before  | After   |
| ------- | ------- |
| ![bug](https://github.com/user-attachments/assets/11eb7c17-dfb3-4830-9e26-110ebdfb8ff7) | ![fixed](https://github.com/user-attachments/assets/89e1b923-f596-4e36-a8eb-fa990ed340d2) |
|![Screenshot 2024-12-17 at 1 24 23 PM](https://github.com/user-attachments/assets/07b72744-c173-447e-90c2-7dcf4ef1334a) | Gen2 regions should no longer be displayed if you don't have permission to access the cluster |
|![Screenshot 2024-12-17 at 10 27 39 PM](https://github.com/user-attachments/assets/ea41e7b7-9472-44e2-8f69-95f19d40a225)|Gen2 regions showing! ![Screenshot 2024-12-17 at 10 27 24 PM](https://github.com/user-attachments/assets/fd8f7216-ec85-4d49-b367-0edcaad376c8)|
|![Screenshot 2024-12-17 at 10 18 24 PM](https://github.com/user-attachments/assets/c2dc7a33-7187-4e67-baec-85deca9e5039)|![Screenshot 2024-12-17 at 10 24 10 PM](https://github.com/user-attachments/assets/43c44384-d38f-47e9-810b-158bebb25dfb)|
## How to test 🧪

### Prerequisites
(How to setup test environment)
- See me and I'll get you added to the engineering account with capabilities

### Reproduction steps
(How to reproduce the issue, if applicable)
- Go to `/object-storage/buckets` in PROD and observe unavailable regions derived from our endpoints are not displayed due to difference in how we formatted the errors.

### Verification steps
(How to verify changes)
- View locally or with preview link and observe unavailable regions should be displaying appropriately.
- After creating a gen2 bucket, try creating an access key in that same region. You should see the bucket appear. 

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>


## As an Author, before moving this PR from Draft to Open, I confirmed ✅


</details>
